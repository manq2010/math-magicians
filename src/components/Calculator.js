import React, { useState } from 'react';
import styled from 'styled-components';
import { colors } from '../theme';
import calculator from '../logic/calculator';

const CalculatorContainer = styled.section`
  display: grid;
  grid-template-columns: repeat(4, 60px);
  grid-template-rows: repeat(6, 50px);
  border: 1px solid ${colors.gray};
  width: fit-content;
  height: fit-content;
  margin: 100px auto;
`;

const Display = styled.article`
  display: flex;
  justify-content: end;
  padding: 0 10px;
  align-items: center;
  border: 1px solid ${colors.gray};
  color: ${colors.white};
  grid-column: 1 / span 4;
  background: ${colors.chrome};
`;

const Clear = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color: ${colors.black};
  grid-column: 1 / 1;
  cursor: pointer;
  &:hover {
    color: ${colors.yellow};
  }
`;

const PlusMinus = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black} ;
  grid-column: 2 / 2;
  cursor: pointer;
  &:hover {
    color: ${colors.yellow};
  }
`;

const Percentage = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black} ;
  grid-column: 3 / 3;
  cursor: pointer;
  &:hover {
    color: ${colors.yellow};
  }
`;

const Divide = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black};
  grid-column: 4 / 4;
  background: ${colors.yellow};
  cursor: pointer;
  &:hover {
    color: ${colors.gray};
  }
`;

const Multiple = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black};
  grid-row: 3 / 3;
  grid-column: 4 / 4;
  background: ${colors.yellow};
  cursor: pointer;
  &:hover {
    color: ${colors.gray};
  }
`;

const Minus = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black};
  grid-row: 4 / 4;
  grid-column: 4 / 4;
  background: ${colors.yellow};
  cursor: pointer;
  &:hover {
    color: ${colors.gray};
  }
`;

const Add = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black};
  grid-row: 5 / 5;
  grid-column: 4 / 4;
  background: ${colors.yellow};
  cursor: pointer;
  &:hover {
    color: ${colors.gray};
  }
`;

const Equal = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black};
  grid-row: 6 / span 1;
  grid-column: 4 / 4;
  background: ${colors.yellow};
  cursor: pointer;
  &:hover {
    color: ${colors.gray};
  }
`;

const Zero = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black};
  grid-row: 6 / span 1;
  grid-column: 1 / span 2;
  cursor: pointer;
  &:hover {
    color: ${colors.yellow};
  }
`;

const Number = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid ${colors.gray};
  color:  ${colors.black};
  cursor: pointer;
  &:hover {
    color: ${colors.yellow};
  }
`;

const Dot = styled.div`
display: flex;
justify-content: center;
border: 1px solid ${colors.gray};
align-items: center;
color:  ${colors.black};
cursor: pointer;
&:hover {
    color: ${colors.yellow};
  }
`;

function Calculator() {
  const [state, setState] = useState({ total: 0, next: 0, operation: '' });

  const onClickHandler = (elem) => {
    const value = elem.getAttribute('value');
    setState(calculator(state, value));
  };

  const { total, next } = state;

  return (
    <CalculatorContainer>
      <Display>
        {next || total || 0}
      </Display>
      <Clear onClick={(e) => onClickHandler(e.target)} value="AC">AC</Clear>
      <PlusMinus onClick={(e) => onClickHandler(e.target)} value="+/-">+/-</PlusMinus>
      <Percentage onClick={(e) => onClickHandler(e.target)} value="%">%</Percentage>
      <Divide onClick={(e) => onClickHandler(e.target)} value="รท">รท</Divide>
      <Multiple onClick={(e) => onClickHandler(e.target)} value="x">*</Multiple>
      <Minus onClick={(e) => onClickHandler(e.target)} value="-">-</Minus>
      <Add onClick={(e) => onClickHandler(e.target)} value="+">+</Add>
      <Equal onClick={(e) => onClickHandler(e.target)} value="=">=</Equal>
      <Zero onClick={(e) => onClickHandler(e.target)} value="0">0</Zero>
      <Number onClick={(e) => onClickHandler(e.target)} value="7">7</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="8">8</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="9">9</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="4">4</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="5">5</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="6">6</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="1">1</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="2">2</Number>
      <Number onClick={(e) => onClickHandler(e.target)} value="3">3</Number>
      <Dot onClick={(e) => onClickHandler(e.target)} value=".">.</Dot>
    </CalculatorContainer>
  );
}

export default Calculator;
