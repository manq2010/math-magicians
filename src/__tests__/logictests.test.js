import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator/Calculator';

describe('Calculator component is rendererd', () => {
  test('Component functions', () => {
    // Arrange
    render(<Calculator />);

    // Act
    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('+'));
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('='));
    const result = screen.getByTestId('total');

    // Assert
    expect(result).toHaveTextContent('15');
  });

  test('matches snapshot', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
