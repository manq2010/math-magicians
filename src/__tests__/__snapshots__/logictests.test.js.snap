// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component is rendererd matches snapshot 1`] = `
<section
  className="sc-bcXHqe kyzHPj"
>
  <article
    className="sc-gswNZR bkLiuQ"
    data-testid="total"
  >
    0
  </article>
  <div
    className="sc-dkrFOg lmfZKB"
    onClick={[Function]}
    value="AC"
  >
    AC
  </div>
  <div
    className="sc-hLBbgP kFHpQd"
    onClick={[Function]}
    value="+/-"
  >
    +/-
  </div>
  <div
    className="sc-eDvSVe dpqsBJ"
    onClick={[Function]}
    value="%"
  >
    %
  </div>
  <div
    className="sc-jSUZER fKsVko"
    onClick={[Function]}
    value="รท"
  >
    รท
  </div>
  <div
    className="sc-gKPRtg gUNOKr"
    onClick={[Function]}
    value="x"
  >
    *
  </div>
  <div
    className="sc-iBYQkv eqtNis"
    onClick={[Function]}
    value="-"
  >
    -
  </div>
  <div
    className="sc-ftTHYK gkeIwO"
    onClick={[Function]}
    value="+"
  >
    +
  </div>
  <div
    className="sc-pyfCe eKEIwg"
    onClick={[Function]}
    value="="
  >
    =
  </div>
  <div
    className="sc-jrcTuL cyaUWH"
    onClick={[Function]}
    value="0"
  >
    0
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="7"
  >
    7
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="8"
  >
    8
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="9"
  >
    9
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="4"
  >
    4
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="5"
  >
    5
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="6"
  >
    6
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="1"
  >
    1
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="2"
  >
    2
  </div>
  <div
    className="sc-kDvujY gCicXC"
    onClick={[Function]}
    value="3"
  >
    3
  </div>
  <div
    className="sc-ipEyDJ dwjWJO"
    onClick={[Function]}
    value="."
  >
    .
  </div>
</section>
`;
